File: frontend/lib/features/timingle/domain/entities/event.dart
import 'package:equatable/equatable.dart';

import '../../../../core/constants/app_constants.dart';

/// 이벤트 참여자
class EventParticipant extends Equatable {
  final int id;
  final String name;
  final String? displayName;
  final String? profileImageUrl;
  final bool confirmed;
  final DateTime? confirmedAt;

  const EventParticipant({
    required this.id,
    required this.name,
    this.displayName,
    this.profileImageUrl,
    this.confirmed = false,
    this.confirmedAt,
  });

  @override
  List<Object?> get props => [
        id,
        name,
        displayName,
        profileImageUrl,
        confirmed,
        confirmedAt,
      ];
}

/// 이벤트 엔티티
class Event extends Equatable {
  final int id;
  final String title;
  final String? description;
  final DateTime startTime;
  final DateTime endTime;
  final String? location;
  final String status;
  final int creatorId;
  final EventParticipant? creator;
  final List<EventParticipant> participants;
  final int? unreadMessages;
  final DateTime createdAt;
  final DateTime? updatedAt;

  const Event({
    required this.id,
    required this.title,
    this.description,
    required this.startTime,
    required this.endTime,
    this.location,
    this.status = AppConstants.statusProposed,
    required this.creatorId,
    this.creator,
    this.participants = const [],
    this.unreadMessages,
    required this.createdAt,
    this.updatedAt,
  });

  @override
  List<Object?> get props => [
        id,
        title,
        description,
        startTime,
        endTime,
        location,
        status,
        creatorId,
        creator,
        participants,
        unreadMessages,
        createdAt,
        updatedAt,
      ];

  /// 상태 체크 헬퍼
  bool get isProposed => status == AppConstants.statusProposed;
  bool get isConfirmed => status == AppConstants.statusConfirmed;
  bool get isCanceled => status == AppConstants.statusCanceled;
  bool get isDone => status == AppConstants.statusDone;

  /// 읽지 않은 메시지 있는지 확인
  bool get hasUnreadMessages => (unreadMessages ?? 0) > 0;

  /// 복사 메서드
  Event copyWith({
    int? id,
    String? title,
    String? description,
    DateTime? startTime,
    DateTime? endTime,
    String? location,
    String? status,
    int? creatorId,
    EventParticipant? creator,
    List<EventParticipant>? participants,
    int? unreadMessages,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Event(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      startTime: startTime ?? this.startTime,
      endTime: endTime ?? this.endTime,
      location: location ?? this.location,
      status: status ?? this.status,
      creatorId: creatorId ?? this.creatorId,
      creator: creator ?? this.creator,
      participants: participants ?? this.participants,
      unreadMessages: unreadMessages ?? this.unreadMessages,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

File: backend/internal/models/event.go
package models

import (
	"time"
)

// EventStatus represents event status types
type EventStatus string

const (
	EventStatusProposed  EventStatus = "PROPOSED"
	EventStatusConfirmed EventStatus = "CONFIRMED"
	EventStatusCanceled  EventStatus = "CANCELED"
	EventStatusDone      EventStatus = "DONE"
)

// Event represents an event/appointment in the system
type Event struct {
	ID               int64       `json:"id" db:"id"`
	Title            string      `json:"title" db:"title"`
	Description      *string     `json:"description,omitempty" db:"description"`
	StartTime        time.Time   `json:"start_time" db:"start_time"`
	EndTime          time.Time   `json:"end_time" db:"end_time"`
	Location         *string     `json:"location,omitempty" db:"location"`
	CreatorID        int64       `json:"creator_id" db:"creator_id"`
	Status           EventStatus `json:"status" db:"status"`
	GoogleCalendarID *string     `json:"google_calendar_id,omitempty" db:"google_calendar_id"` // Google Calendar 연동 ID
	CreatedAt        time.Time   `json:"created_at" db:"created_at"`
	UpdatedAt        time.Time   `json:"updated_at" db:"updated_at"`
}

// EventParticipant represents a participant in an event
type EventParticipant struct {
	EventID     int64      `json:"event_id" db:"event_id"`
	UserID      int64      `json:"user_id" db:"user_id"`
	Confirmed   bool       `json:"confirmed" db:"confirmed"`
	ConfirmedAt *time.Time `json:"confirmed_at,omitempty" db:"confirmed_at"`
}

// CreateEventRequest represents event creation request
type CreateEventRequest struct {
	Title         string    `json:"title" binding:"required"`
	Description   *string   `json:"description,omitempty"`
	StartTime     time.Time `json:"start_time" binding:"required"`
	EndTime       time.Time `json:"end_time" binding:"required"`
	Location      *string   `json:"location,omitempty"`
	ParticipantIDs []int64  `json:"participant_ids,omitempty"`
}

// UpdateEventRequest represents event update request
type UpdateEventRequest struct {
	Title       *string     `json:"title,omitempty"`
	Description *string     `json:"description,omitempty"`
	StartTime   *time.Time  `json:"start_time,omitempty"`
	EndTime     *time.Time  `json:"end_time,omitempty"`
	Location    *string     `json:"location,omitempty"`
	Status      *EventStatus `json:"status,omitempty"`
}

// EventResponse represents event data in API responses
type EventResponse struct {
	ID           int64              `json:"id"`
	Title        string             `json:"title"`
	Description  *string            `json:"description,omitempty"`
	StartTime    time.Time          `json:"start_time"`
	EndTime      time.Time          `json:"end_time"`
	Location     *string            `json:"location,omitempty"`
	Creator      *UserResponse      `json:"creator"`
	Participants []*UserResponse    `json:"participants,omitempty"`
	Status       EventStatus        `json:"status"`
	CreatedAt    time.Time          `json:"created_at"`
	UpdatedAt    time.Time          `json:"updated_at"`
}

// EventWithParticipants represents an event with its participants
type EventWithParticipants struct {
	Event        *Event
	Creator      *User
	Participants []*User
}

// ToEventResponse converts EventWithParticipants to EventResponse
func (e *EventWithParticipants) ToEventResponse() *EventResponse {
	response := &EventResponse{
		ID:          e.Event.ID,
		Title:       e.Event.Title,
		Description: e.Event.Description,
		StartTime:   e.Event.StartTime,
		EndTime:     e.Event.EndTime,
		Location:    e.Event.Location,
		Status:      e.Event.Status,
		CreatedAt:   e.Event.CreatedAt,
		UpdatedAt:   e.Event.UpdatedAt,
	}

	if e.Creator != nil {
		response.Creator = e.Creator.ToUserResponse()
	}

	if len(e.Participants) > 0 {
		response.Participants = make([]*UserResponse, len(e.Participants))
		for i, participant := range e.Participants {
			response.Participants[i] = participant.ToUserResponse()
		}
	}

	return response
}
